PhpAmqpLib/Wire/IO/AbstractIO.php

```diff
@@ -40,6 +40,9 @@ abstract class AbstractIO
     /** @var int|float */
     protected $last_write;
 
+    /** @var int|float */
+    protected $last_heartbeat;
```

Comment:
last heartbeat will be same as last write?

Reply:
last_write is updated after the writing is done, while last_heartbeat will be updated before writing. This is needed due to the posibility of interrupt happening during `write_heartbeat,` when last_write has not been updated yet.

Reply:
Isn't it better to use `$conection->writing` to prevent whole heartbeat check comming from outside?

Reply:
`check_heartbeat` is called from IO::read as well, when `$connection->writing` is set to _false_

Instead of using `last_heartbeat`, I could add a `writing_heartbeat` state variable in IO and call `isWritingHeartbeat` from connection during `isWriting` as an additional condition. Not sure if that is a better solution, though

Reply:
Well yes, but IO::read is internal call before read, is not possible to interupt writeable frame there.
You add `$this->connection->isWriting()` protection to PCNTLHeartbeatSender, but in fact this protection is required for all external calls, i.e. tick based heartbeat checkers.
Also, time difference between last_write and last_heartbeat will be insignificant, hardly it will make any real difference for heartbeats longer than a few seconds.